### PR TITLE
Add Linked Expressions to Nuke Gizmos

### PR DESCRIPTION
**Pull request by kateforsberg13**
Updating to allow for expression linking.
Adds `Link` buttons to the nuke nodes, which can expression link to string inputs.
image
Screenshot 2026-07-28 at 2 37 22 PM
Screenshot 2026-07-28 at 2 37 38 PM
Screenshot 2026-07-28 at 2 37 53 PM

### CODE REVIEW

File: publish_gizmo/nuke_gizmo_builder.py [hunk 1]
```diff
@@ -114,23 +140,119 @@ def _label(name: str) -> str:
...
-        switch = nuke.toNode("SwitchOutput")
-        if switch:
-            switch["which"].setValue(int(n["active_output"].getValue()))
+        _gt_switch = nuke.toNode("SwitchOutput")
+        if _gt_switch:
+            _gt_switch["which"].setValue(int(_gt_n["active_output"].getValue()))
     finally:
-        n.end()
+        _gt_n.end()
+el"""
```
Comment (hutchinson) [inline publish_gizmo/nuke_gizmo_builder.py]: Great work restructuring the knobChanged callback to handle both the rename/unlink upkeep and the active_output switching — the consolidated approach is cleaner than two separate callbacks.

One concern worth addressing before merge: `active_output_block` ends with the bare characters `el`, relying on the f-string below concatenating `{prefix}if` immediately after to form `elif`. This works today, but it's a trap for future editors — any whitespace, blank line, or reordering between those two strings produces a syntax error that only surfaces at runtime inside Nuke, with no static analysis or test catching it.

A more robust approach would be to always emit both branches and guard the `active_output` body with `if nuke.toNode("SwitchOutput"):` — it's a no-op when the node doesn't exist, and removes the conditional code generation entirely. Alternatively, build the two branches as complete strings and join them with a proper `elif`.

**Blocking** — the failure mode is silent and runtime-only.

File: publish_gizmo/run_button.py [hunk 2]
```diff
@@ -64,6 +64,34 @@ def _build_child_env(companion: str, base_env: dict) -> dict:
...
+    node context so ``this`` resolves to the gizmo instance. Values that aren't
+    valid TCL (e.g. JSON arrays) raise and fall back to the raw text.
+    """
+    expr_knob = n.knob("_gt_expr_" + knob_name)
+    if expr_knob is not None and expr_knob.getText():
+        try:
+            expanded = expr_knob.evaluate()
+        except Exception:  # noqa: BLE001
+            expanded = None
+        if expanded:
```
Comment (hutchinson) [inline publish_gizmo/run_button.py]: Nice docstring on `_expand_tcl` — the two-path resolution (stored expression → hand-typed brackets → raw fallback) is exactly right.

One edge case worth fixing: both falsy guards treat `""` as a failure and fall through to `raw`:

```python
if expanded:          # "" is falsy — falls through
    return expanded
...
return expanded if expanded else raw   # same issue
```

If a linked expression legitimately resolves to an empty string (e.g. `[value root.name]` before any Nuke script is open), `output_dir` silently uses whatever stale text is in the display field rather than the empty string the expression produced. Since `if output_dir:` immediately follows this call, passing `""` through would be correct.

Recommend `is not None` instead of truthiness:

```python
if expanded is not None:
    return expanded
```

**Blocking** — silent wrong-value substitution on legitimate empty expansions.

File: tests/unit/test_run_button_expand_tcl.py [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+        self._expr_knobs = expr_knobs or {}
+
+    def fullName(self) -> str:  # noqa: N802
+        return self._full_name
+
+    def knob(self, name: str):
+        return self._expr_knobs.get(name)
+
+
+def _load_expand_tcl(fake_nuke: _FakeNuke):
```
Comment (hutchinson) [inline tests/unit/test_run_button_expand_tcl.py]: The fake-nuke harness here is inventive and the test cases are thorough — good coverage of all the edge cases.

The `_load_expand_tcl` extraction approach is worth a note: scanning `run_button.py` source for `def _expand_tcl(` at column 0 and collecting indented lines will break silently if the function is decorated, moved, or a module-level comment lands between it and the next definition.

The root cause is that `run_button.py` is `exec`'d rather than imported. A cleaner fix would be to move `_expand_tcl` into a small importable helper (e.g. `publish_gizmo/tcl_utils.py`) that `run_button.py` imports at runtime — the tests could then import it normally with no source-parsing, and `run_button.py` stays as the exec'd entry point. Not blocking, but worth considering for maintainability.

File: tests/unit/test_nuke_gizmo_builder.py [hunk 4]
```diff
@@ -225,4 +225,162 @@ def test_non_media_output_does_not_produce_read_node(self) -> None:
...
+        knob_line = next(line for line in gizmo.splitlines() if "addUserKnob {1 folder_path" in line)
+        assert ' t "' in knob_line
+        assert r"\[value this.name]" in knob_line
+
+    def test_tooltips_never_contain_unescaped_brackets(self) -> None:
+        gizmo = _generate_gizmo({"folder_path": {"type": "str", "default_value": ""}})
+        for line in gizmo.splitlines():
+            # PyScript button lines carry raw Python in T "..." where brackets are
+            # brace-quoted and safe; only value-knob tooltips need escaping.
+            if ' t "' in line and "addUserKnob {22" not in line:
```
Comment (hutchinson) [inline tests/unit/test_nuke_gizmo_builder.py]: The tooltip escaping tests are a great addition. One small gap: the exclusion on line 266 skips all `{22}` (PyScript knob) lines, including their `t "tooltip"` attributes. `_COPY_LINK_TOOLTIP` contains `[value ...]` — which `_tcl_escape_literal` does escape correctly — but this test never verifies it. If someone later edits that string and adds an unescaped bracket, nothing catches it.

The comment's reasoning ("brackets are brace-quoted and safe") is also slightly off — the `T "..."` script body isn't brace-quoted; brackets there just have no TCL significance because Nuke doesn't perform substitution inside the script attribute.

Suggest either removing the `{22}` exclusion and checking only the `t "..."` substring of each line, or adding a dedicated test asserting the Copy Link button's tooltip line contains no unescaped `[`.

File: publish_gizmo/nuke_gizmo_builder.py [hunk 5]
```diff
@@ -114,23 +140,119 @@ def _label(name: str) -> str:
...
+    prefix = active_output_block if has_multiple_media_outputs else ""
+    return f"""\
+_gt_k = nuke.thisKnob()
+_gt_n = nuke.thisNode()
+_gt_kn = _gt_k.name()
+{prefix}if _gt_kn == "name":
+    for _gt_name in list(_gt_n.knobs()):
+        if _gt_name.startswith("_gt_expr_"):
```
Comment (hutchinson) [inline publish_gizmo/nuke_gizmo_builder.py]: Minor maintainability note: the prefix `"_gt_expr_"` appears verbatim in the generated `knobChanged` code here, in `_write_link_button`, and in `_expand_tcl` in `run_button.py`. If it ever needs changing it breaks in three places with no compile-time signal.

A module-level constant in `publish_gizmo/constants.py` (e.g. `GT_EXPR_PREFIX = "_gt_expr_"`) would tie them together. Not blocking — just worth a follow-up.

File: publish_gizmo/nuke_gizmo_builder.py [hunk 6]
```diff
@@ -72,6 +72,32 @@
...
+    "Link this field to the gizmo name, script folder, script name, frame, or any node.knob. "
+    "The field shows the linked value; editing it by hand removes the link."
+)
+_COPY_LINK_TOOLTIP = (
+    "Copy a [value ...] expression for this output to the clipboard. "
+    "Select another node first to expression-link one of its knobs to this output directly."
+)
+
+
+def _output_tooltip(knob_name: str) -> str:
```
Comment (hutchinson) [inline publish_gizmo/nuke_gizmo_builder.py:94]: The output tooltip is a nice UX touch — giving operators a ready-made expression to copy is exactly the right thing to do here.

One small note: `knob_name` at this point already carries the `_out` suffix (e.g. `caption_out`), so the tooltip reads "e.g. `[value <this node's name>.caption_out]`" while the knob is labelled "Caption" in the panel. The expression is correct — that's what the user needs to type — but the mismatch with the visible label could read as a typo. A concrete worked example or a brief note that the suffix is intentional would make it clearer. Totally optional.

File: publish_gizmo/nuke_gizmo_builder.py [hunk 7]
```diff
@@ -114,23 +140,119 @@ def _label(name: str) -> str:
...
+                if _gt_v:
+                    _gt_n[_gt_target].setValue(_gt_v)
+elif not _gt_kn.startswith("_gt_") and _gt_n.knob("_gt_expr_" + _gt_kn):
+    _gt_ek = _gt_n["_gt_expr_" + _gt_kn]
+    if _gt_ek.getText():
+        try:
+            _gt_cur = _gt_ek.evaluate()
+        except Exception:
+            _gt_cur = None
+        if _gt_k.getText() != _gt_cur:
```
Comment (hutchinson) [inline publish_gizmo/nuke_gizmo_builder.py:193]: The unlink detection logic is solid. One known limitation worth capturing in a comment: if the expression evaluates to the same text the user then types manually (e.g. the link resolves to `"Gizmo1"` and the user types `"Gizmo1"`), the link won't be cleared because the values are equal. This is an acceptable tradeoff for a value-comparison approach, but a future maintainer might read it as a bug without that context — a short comment here would help.

File: publish_gizmo/gizmo_writer.py [hunk 8]
```diff
@@ -180,16 +197,19 @@ def add_enumeration_knob(self, knob_name: str, label: str, choices: list, defaul
...
 
         ``python_code`` is plain Python; this method handles TCL escaping.
         """
         self._knobs.append((NukeKnobType.PYSCRIPT, knob_name))
         escaped = _tcl_escape(python_code)
+        tooltip_part = f' t "{_tcl_escape_literal(tooltip)}"' if tooltip else ""
         flag_suffix = f" {flags}" if flags else ""
         self._lines.append(
-            f' addUserKnob {{{NukeKnobType.PYSCRIPT} {knob_name} l "{label}" T "{escaped}"{flag_suffix}}}'
+            f' addUserKnob {{{NukeKnobType.PYSCRIPT} {knob_name} l "{label}"{tooltip_part} T "{escaped}"{flag_suffix}}}'
```
Comment (hutchinson) [inline publish_gizmo/gizmo_writer.py:212]: The `t "tooltip"` addition looks correct. One thing worth a quick manual check in Nuke: the generated attribute order is now `l "label" t "tooltip" T "code"`, whereas the previous pattern had `T` immediately after `l`. Nuke's `addUserKnob` parser is generally order-insensitive for named attributes, so this is likely fine — just worth confirming the tooltip actually renders on `{22}` PyScript knobs in a real session, since the unit tests don't load the gizmo in Nuke.

**Review by hutchinson — CHANGES_REQUESTED** [review]
This review was produced with AI assistance (Claude Code) and verified by a human reviewer before submission.

The implementation approach is correct and well-thought-out — the hidden companion-knob pattern and `_expand_tcl` two-path resolution are the right solutions given how Nuke string knobs work. The test coverage is strong. Eight items below, two of which (correctness issues) are blocking; the rest are advisory.

File: publish_gizmo/nuke_gizmo_builder.py
Comment (hutchinson) [inline publish_gizmo/nuke_gizmo_builder.py:1]: Perhaps we can log an issue to come back later and refactor this file to be written interms of the Nuke Python API directly. We have some of the infrastructure required to do this already in this repo.

Comment (kateforsberg13) [reply publish_gizmo/nuke_gizmo_builder.py:1]: yes, valid. We may have an issue for this already, but the thought here was that it's much faster on publish when you don't have to load `nuke`. However, you may be right that we've gotten to the point where the accuracy is worth the speed tradeoff.

**Review by hutchinson — APPROVED** [review]
